include/boost/beast/core/impl/buffers_adaptor.hpp

```diff
@@ -601,6 +437,329 @@ consume(std::size_t n) noexcept
+        last_ = std::next(current);
+        last_size_ = n;
+    }
+    else
+    {
+        n -= max_elem;
+        ++current;
+    }
+
+    for (;;)
+    {
+        max_elem = current->size();
+        if (last_element(current))
+        {
+            last_size_ = (std::min)(n, last_size_);
+            return;
+        }
+        else if (max_elem < n)
+        {
+            n -= max_elem;
+            ++current;
+        }
+        else
+        {
+            last_size_ = n;
+            last_ = std::next(current);
+            return;
+        }
+    }
+}
+
+
+template<class MutableBufferSequence>
+template<bool isMutable>
+auto
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+begin() const -> iterator
+{
+    return iterator(this, first_);
+}
+
+template<class MutableBufferSequence>
+template<bool isMutable>
+auto
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+end() const -> iterator
+{
+    return iterator(this, last_);
+}
+
+template<class MutableBufferSequence>
+template<bool isMutable>
+auto
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+get_buffer(iter_type it) const
+-> value_type
+{
+    auto result = *it;
+
+    if (it == first_)
+        result += first_offset_;
+
+    if (std::next(it) == last_)
+        detail::truncate(result, last_size_);
+
+    return result;
+}
+
+template<class MutableBufferSequence>
+template<bool isMutable>
+auto
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+first() const -> iter_type
+{
+    return first_;
+}
+
+template<class MutableBufferSequence>
+template<bool isMutable>
+auto
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+last() const -> iter_type
+{
+    return last_;
+}
+
+template<class MutableBufferSequence>
+template<bool isMutable>
+auto
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+first_offset() const -> std::size_t
+{
+    return first_offset_;
+}
+
+template<class MutableBufferSequence>
+template<bool isMutable>
+auto
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+last_size() const -> std::size_t
+{
+    return last_size_;
+}
+
+// -------------------------------------------------------------------------
+// buffers_adaptor::subrange::iterator
+
+template<class MutableBufferSequence>
+template<bool isMutable>
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+iterator::
+iterator()
+    : parent_(nullptr)
+    , it_()
+{
+}
+
+template<class MutableBufferSequence>
+template<bool isMutable>
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+iterator::
+iterator(
+    buffers_adaptor<MutableBufferSequence>::subrange<isMutable> const *parent,
+    iter_type it)
+    : parent_(parent)
+    , it_(it)
+{
+}
+
+template<class MutableBufferSequence>
+template<bool isMutable>
+auto
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+iterator::
+operator*() const
+-> value_type
+{
+    return parent_->get_buffer(it_);
```

Comment:
There's no need for `get_buffer` since it is only called from this place, might as well just put the code here.

Reply:
This boils down to our fundamentally different views on nested classes. I disagree, but will change it.